server.py

```diff
@@ -414,24 +413,34 @@ def asyncFinishThread(server: 'Server'):
 		errorTracker.put(server.get_error_pk(), (False, err))
 		return
 
-	if checkAddress:
+	if server.ip == server.address:
+		server.verifyLevel = 3
+	else:
 		addresses = set(data[4][0] for data in info)
 		have_v4 = any(d[0] == socket.AF_INET for d in info)
 		have_v6 = any(d[0] == socket.AF_INET6 for d in info)
 		if server.ip in addresses:
-			pass
+			server.verifyLevel = 3
 		elif (":" in server.ip and not have_v6) or ("." in server.ip and not have_v4):
 			# If the client is ipv6 and there is no ipv6 on the domain (or the inverse)
 			# then the check cannot possibly ever succeed.
 			# Because this often happens accidentally just tolerate it.
-			pass
+			server.verifyLevel = 2
 		else:
 			err = "Requester IP %s does not match host %s" % (server.ip, server.address)
 			if isDomain(server.address):
 				err += " (valid: %s)" % " ".join(addresses)
 			app.logger.warning(err)
 			# handle as warning
 			errorTracker.put(server.get_error_pk(), (True, err))
+			server.verifyLevel = 1
+
+	# do this here since verifyLevel is now set
+	if serverList.checkDuplicate(server):
```

Comment:
This only returns true if it's a duplicate of a different server and that server is more trusted, so maybe this should be named something like `isImpersonation` instead.

Reply:
calling this "impersonation" feels heavy-handed when there are innocent reasons for this to happen